Design.md

```diff
@@ -309,3 +332,29 @@ Get the platform usage of global (without query parameter), tenant (with tenant_
         * `key` : `offset` and `value` : `marker`
     1. `AssumeRoleCache` for all the `User` and `S3 Credential` APIs
         * `key` : `RoleArn` and `value` : `Credentials`
+
+## SecretKey Encryption Strategy
+
+1. For encrypting/ decrypting the `secret-keys` a `cipher` algorithm, and a `secret-password` will be used. These properties will be provided in `crypto.properties` config file.
+1. A `key-id` property will be provided for each `secret-password` in the `crypto.properties` config file to maintain the version of the key and will be used in the Key rotation.
+1. Meta-information will be encoded with base64 and added at the beginning of the encrypted string. This meta information will include the `key-id` of the `secret-password` used.
+1. All the secret-passwords in the `crypto.properties` config file will be base64 encoded. These secret-passwords will be decoded using base64 before encrypting the `secret-keys`
+1. Initially only `AES256GCM` cipher algorithm will be supported. (`secret-box` algorithm is suggested to be implemented for future releases)
+1. Adaptability to change the cipher algorithm will be provided.
+1. Example `crypto.properties`
+    ```yaml
+    keys:
+      - id: 2
+        cipher: Secretbox
+        secretKey: dGhpc2lzYXJlYWxseWxvbmdhbmRzdHJvbmdrZXk=
+    ```
+1. **Key Rotation:**
+    * Key rotation is used to replace a key with another one once the original key is compromised or too old to use.
+    * A new API will be provided with the `OSIS` application to do the key rotation of all the encrypted `secret-key` data on Redis database. 
+        1. Key rotation API will use the meta-information of the encrypted data to identify the `key-id` used.
+        1. Each `secret-key` will be decrypted using the old cipher algorithm (identified using the `key-id`).
+        1. `secret-key` value will be re-encrypted using the latest cipher algorithm and updated on Redis.
+        1. This API will be idempotent.
+        1. This API need to be re-run until it returns success which means all the `secret-key` values on Redis are re-encrypted. 
+    * `config.properties` will be updated with the latest keys.
+    * A `refresh` API will be provided with the `OSIS` application to reload the latest keys from the `config.properties` file.
```

Comment:
Also, the application should re-read its config file when it finds a key-id in an entry retrieved from the datastore that it hasn't seen before.

Reply:
Fixed in 380d8acd7a40b01860989f28d0eb985b95fc5a61